### PR TITLE
De duplicate PCR profile branches

### PR DESCRIPTION
**Pull request by chrisccoulson**
AddEFISecureBootPolicyProfile contains logic for de-duplicating secure boot policy configurations, but this logic really belongs in PCRProtectionProfile. This PR adds the logic there, and a follow-up will clean up the interaction between AddEFISecureBootPolicyProfile and PCRProtectionProfile.

To allow this, this PR changes the interaction between PCRProtectionProfile and the core policy logic. Previously, PCRProtectionProfile would compute sets of PCR values for each branch, and then computeDynamicPolicy would compute a PCR selection and list of PCR digests from this. Now, PCRProtectionProfile computes the PCR selection and PCR digests directly, which makes it easier to remove duplicate branches.

### CODE REVIEW

**Comment by pedronis** [comment]
@chrisccoulson this probably merits a review also from the security team

**Comment by chrisccoulson** [comment]
> @chrisccoulson this probably merits a review also from the security team

Possibly, although I'm not sure who best to look at it. The changes here are largely mechanical (computing PCR digests from sets of PCR values is just moved inside of PCRProtectionProfile).

@setharnold perhaps?

File: policy.go [hunk 1]
```diff
@@ -50,8 +49,10 @@ type dynamicPolicyComputeParams struct {
 	// signAlg is the digest algorithm for the signature used to authorize the generated dynamic authorization policy. It must
 	// match the name algorithm of the public part of key that will be loaded in to the TPM for verification.
 	signAlg              tpm2.HashAlgorithmId
-	pcrValues            []tpm2.PCRValues // Approved PCR digests
-	policyCountIndexName tpm2.Name        // Name of the NV index used for revoking authorization policies
+	pcrs                 tpm2.PCRSelectionList // PCR selection
+	pcrDigests           tpm2.DigestList       // Approved PCR digests
+	pcrValues            []tpm2.PCRValues      // Approved PCR digests
```
[the remark this answers is not in the document]
Comment (chrisccoulson) [reply policy.go:54]: This field actually should be deleted, as it's unused.

File: pcr_profile_test.go [hunk 2]
```diff
@@ -249,22 +261,64 @@ func TestPCRProtectionProfile(t *testing.T) {
 				},
 			},
 		},
+		{
+			// Verify that other PCR digest algorithms work
+			desc: "SHA1",
```
[the remark this answers is not in the document]
Comment (chrisccoulson) [reply pcr_profile_test.go:266]: SHA1 is the algorithm used for computing the PCR digests from a set of PCR values. SHA256 in this test is used to select the PCR bank to read values from.

File: pcr_profile_test.go [hunk 3]
```diff
@@ -283,14 +337,19 @@ func TestPCRProtectionProfileAddValueFromTPM(t *testing.T) {
...
 	p := NewPCRProtectionProfile().AddPCRValueFromTPM(tpm2.HashAlgorithmSHA256, 7)
-	values, err := p.ComputePCRValues(tpm.TPMContext)
+	pcrs, digests, err := p.ComputePCRDigests(tpm.TPMContext, tpm2.HashAlgorithmSHA256)
 	if err != nil {
-		t.Fatalf("ComputePCRValues failed: %v", err)
+		t.Fatalf("ComputePCRDigests failed: %v", err)
+	}
+
+	if !pcrs.Equal(tpm2.PCRSelectionList{{Hash: tpm2.HashAlgorithmSHA256, Select: []int{7}}}) {
+		t.Errorf("ComputePCRDigests returned the wrong selection")
```
[the remark this answers is not in the document]
Comment (chrisccoulson) [reply pcr_profile_test.go:346]: Fatalf stops terminates the test, and Errorf causes the test to fail but allows execution to continue. The 7 is just the PCR handle used in this test.

**Comment by pedronis** [comment]
@chrisccoulson btw the conflicts with master involve structures that are now versioned there, please ping me when that is solved and I will look at this

File: pcr_profile.go [hunk 4]
```diff
@@ -193,3 +195,42 @@ func (p *PCRProtectionProfile) computePCRValues(tpm *tpm2.TPMContext, values pcr
...
+	var pcrDigests tpm2.DigestList
+	for _, v := range values {
+		p, digest, _ := tpm2.ComputePCRDigestSimple(alg, v)
+		if !p.Equal(pcrs) {
+			return nil, nil, errors.New("not all branches contain values for the same sets of PCRs")
+		}
+		pcrDigests = append(pcrDigests, digest)
+	}
+
+	var filteredPcrDigests tpm2.DigestList
```
Comment (pedronis) [inline pcr_profile.go:220]: this more like unique than filtered, no?

File: pcr_profile.go [hunk 5]
```diff
@@ -193,3 +195,42 @@ func (p *PCRProtectionProfile) computePCRValues(tpm *tpm2.TPMContext, values pcr
 
 	return values, nil
 }
+
+// computePCRDigests computes a PCR selection and list of PCR digests from this PCRProtectionProfile.
```
Comment (pedronis) [inline pcr_profile.go:199]: the doc should mention that the digest are unique/de-duplicated?

**Pull request merged**